### PR TITLE
feat(customer): implement Customer service interface

### PR DESCRIPTION
**Pull request by mkucmus**
Related issue: #63

### CODE REVIEW

File: packages/shopware-6-client/__tests__/services/customerService.spec.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      }
+    });
+  });
+  describe("logout", () => {
+    it("should return no data and throw no exceotion (204)", async () => {
+      try {
+        const contextTokenResult = await CustomerService.login(
+          loginCredentials
+        );
+        const result = await CustomerService.logout(
```
Comment (mkucmus) [inline packages/shopware-6-client/__tests__/services/customerService.spec.ts]: **TODO**

File: packages/shopware-6-client/src/services/customerService.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * @description End up the session
+ */
+const logout = async function(contextToken?: string): Promise<null> {
+  const resp = await axios.post(
+    `${config.endpoint}${CUSTOMER_ENDPOINT}/logout`,
+    null,
+    {
+      headers: {
+        /** TODO: move into different layer if created */
+        "sw-context-token": contextToken
```
Comment (mkucmus) [inline packages/shopware-6-client/src/services/customerService.ts]: **TODO**

File: packages/shopware-6-client/__tests__/services/customerService.spec.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import { CustomerService } from "../../src/index";
+
+const generateUniqueEmail = (): string => {
+  const currentTime = Date.now();
+  const email = `mkucmus+${currentTime}@divante.com`;
+  return email;
+};
+
+const basicCustomerData = {
```
Comment (mkucmus) [inline packages/shopware-6-client/__tests__/services/customerService.spec.ts]: @patzick do we have here a good place to apply faker generated data?

**Comment by mkucmus** [comment]
I've covered all the methods which are available through the REST API. 

@patzick @rmakara I've got this feeling that I should have changed some customer methods and prefix them with "customer" word if we still want to expose all of them in one entrypoint (src/index.js)

File: packages/shopware-6-client/src/services/customerService.ts [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+): Promise<CustomerLoginResponse> {
+  const resp = await apiService.post(`${getCustomerEndpoint()}/login`, params);
+  return resp.data;
+};
+
+/**
+ * @description End up the session
+ */
+export const logout = async function(contextToken?: string): Promise<null> {
+  const resp = await apiService.post(`${getCustomerEndpoint()}/logout`, null, {
```
Comment (patzick) [inline packages/shopware-6-client/src/services/customerService.ts]: we need to have these steps here

1. if contextToken is provided, then call `update({contextToken})`
2. call api here
3. call `update({contextToken: ''})`

don't return anything as it's a void method

File: packages/shopware-6-client/src/services/customerService.ts [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    }
+  });
+  return resp.data;
+};
+
+/**
+ * @description End up the session
+ */
+export const getCustomer = async function(
+  contextToken: string
```
Comment (patzick) [inline packages/shopware-6-client/src/services/customerService.ts]: we don't need to pass contextToken anywheere else than `setup` or `update`
```suggestion
```

File: packages/shopware-6-client/__tests__/services/CustomerService/deleteCustomerAddress.spec.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  it("returns no data if successfully deleted", async () => {
+    mockedAxios.delete.mockResolvedValue(null);
+
+    await deleteCustomerAddress("2bbb89dfa4664bc681e80b37eaa80fb7");
+    expect(mockedAxios.delete).toBeCalledTimes(1);
+    expect(mockedAxios.delete).toBeCalledWith(
+      getCustomerAddressEndpoint("2bbb89dfa4664bc681e80b37eaa80fb7")
+    );
+  });
```
Comment (rmakara) [inline packages/shopware-6-client/__tests__/services/CustomerService/deleteCustomerAddress.spec.ts]: Possible to add test for following behavior:

`Customer address with id \"324af469318f46b68e0fe69d77ef15fb\" is a default address and cannot be deleted.`

However, it is not necessary for now.

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddress.spec.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    jest.resetAllMocks();
+
+    contextToken = random.uuid();
+
+    update({ contextToken });
+  });
+  afterEach(() => {
+    // expect(
+    //   mockedAxios.defaults.headers.common["sw-context-token"]
+    // ).toEqual(contextToken);
```
Comment (rmakara) [inline packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddress.spec.ts]: We may delete that comment. : ) Probably it is one of the tested scenarios during remote pair programming, which is no longer needed. : )

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomer.spec.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    expect(mockedAxios.get).toBeCalledTimes(1);
+    expect(mockedAxios.get).toBeCalledWith(getCustomerEndpoint());
+    expect(result.id).toEqual("c370eb5cd1df4d4dbcc78f055b693e79");
+  });
+
+  it("should an current customer's data - using incorrect token", async () => {
+    mockedAxios.get.mockRejectedValue(new Error());
+    expect(getCustomer()).rejects.toThrow();
+    expect(mockedAxios.get).toBeCalledTimes(1);
+    expect(mockedAxios.get).toBeCalledWith(getCustomerEndpoint());
```
Comment (rmakara) [inline packages/shopware-6-client/__tests__/services/CustomerService/getCustomer.spec.ts]: I am not sure if it's needed but when testing a behavior of some methods, that use incorrect token we may overwrite the correct token in config.contextToken with some invalid one. 

Right now we are testing response from the API, which is mocked to Error() but we are not sure about behavior of `getCustomer()` method. If that would be a black box method then we could assume there could be some validation, that will not even call `get` http method if token does not meet some format requirements.

Not a required change but from my perspective erasing the correct token would be a good step. :) Up to you to decide. :)

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddresses.spec.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    afterEach(() => {
+      expect(config.contextToken).toEqual(contextToken);
+    });
+  });
+
+  it("should return object of addresses", async () => {
+    mockedAxios.get.mockResolvedValueOnce({ data: { data: {} } });
+    const result = await getCustomerAddresses();
+    expect(mockedAxios.get).toBeCalledTimes(1);
+    expect(mockedAxios.get).toBeCalledWith(`${getCustomerEndpoint()}/address`);
```
Comment (rmakara) [inline packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddresses.spec.ts]: There is `getCustomerAddressEndpoint()` in `endpoints.ts`, that we may use. :)

File: packages/shopware-6-client/src/services/customerService.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  const contextToken = resp.data["sw-context-token"];
+  update({ contextToken });
+  return { contextToken };
+}
+
+/**
+ * End up the session
+ */
+export async function logout(): Promise<void> {
+  await apiService.post(`${getCustomerEndpoint()}/logout`);
```
Comment (rmakara) [inline packages/shopware-6-client/src/services/customerService.ts]: Possible to use `getCustomerLogoutEndpoint()`. :)

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddresses.spec.ts (hunk 9 above)
Comment (mkucmus) [reply packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddresses.spec.ts]: you are right, I replaced it with the existing method call

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomer.spec.ts (hunk 8 above)
Comment (mkucmus) [reply packages/shopware-6-client/__tests__/services/CustomerService/getCustomer.spec.ts]: You are totally right, I just left the 400 and promis rejection without any extra info to just break the logic from getCustomer() and end up with unhandled rejection. We will prepare the consistent error throwing/handling in the next few iterations once we discover all the cases :+1:

File: packages/shopware-6-client/src/services/customerService.ts (hunk 10 above)
Comment (mkucmus) [reply packages/shopware-6-client/src/services/customerService.ts]: sure, fixed

File: packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddress.spec.ts (hunk 7 above)
Comment (mkucmus) [reply packages/shopware-6-client/__tests__/services/CustomerService/getCustomerAddress.spec.ts]: removed

**Comment by mkucmus** [comment]
I'm proud of the hard work we put into this PR. Thanks for your support guys!

**Pull request merged**